src/client/QXmppHttpUploadManager.cpp

```diff
@@ -0,0 +1,372 @@
+///
+
+QXmppHttpUpload::~QXmppHttpUpload() = default;
+
+///
+/// \class QXmppHttpUploadManager
+///
+/// The upload manager allows to upload a file to a server via HTTP File Upload (XEP-363).
+/// This can be used for sending files to other users.
+///
+/// You can enable it using
+/// \code
+/// client.addNewExtension<QXmppUploadRequestManager>();
+/// auto *uploadManager = client.addNewExtension<QXmppHttpUploadManager>();
+/// \endcode
+///
+
+///
+/// \typedef QXmppHttpUpload::Result
+///
+/// Represents the result of an upload.
+/// It can either be a an url, to the uploaded file,
+/// a QXmppHttpUpload::Cancelled unit struct, or an error as QXmppError.
+///
+
+///
+/// Returns the current progress of the upload as a floating point number between 0 and 1.
+///
+float QXmppHttpUpload::progress() const
+{
+    if (d->bytesTotal > 0) {
+        if (d->bytesSent > d->bytesTotal) {
+            return 1;
+        }
+        return float(d->bytesSent) / d->bytesTotal;
+    }
+    return 0;
+}
+
+///
+/// The number of bytes sent so far.
+///
+quint64 QXmppHttpUpload::bytesSent() const
+{
+    return d->bytesSent;
+}
+
+///
+/// The number of bytes that need to be sent in total to complete the upload.
+///
+quint64 QXmppHttpUpload::bytesTotal() const
+{
+    return d->bytesTotal;
+}
+
+///
+/// Cancel the upload
+///
+void QXmppHttpUpload::cancel()
+{
+    d->cancelled = true;
+    if (d->reply) {
+        d->reply->abort();
+    }
+}
+
+///
+/// Returns whether the upload is already finished.
+///
+bool QXmppHttpUpload::isFinished() const
+{
+    return d->finished;
+}
+
+///
+/// If the upload has already finished, returns the result of the upload as QXmppHttpUpload::Result,
+/// otherwise returns an empty std::optional.
+///
+std::optional<QXmppHttpUpload::Result> QXmppHttpUpload::result() const
+{
+    if (d->finished) {
+        return d->result();
+    }
+    return {};
+}
+
+QXmppHttpUpload::QXmppHttpUpload()
+    : d(std::make_unique<QXmppHttpUploadPrivate>(this))
+{
+}
+
+struct QXmppHttpUploadManagerPrivate
+{
+    explicit QXmppHttpUploadManagerPrivate(QNetworkAccessManager *netManager)
+        : netManager(netManager)
+    {
+    }
+
+    QNetworkAccessManager *netManager;
+};
+
+///
+/// Constructor
+///
+/// Creates and uses a new network access manager.
+///
+QXmppHttpUploadManager::QXmppHttpUploadManager()
+    : d(std::make_unique<QXmppHttpUploadManagerPrivate>(new QNetworkAccessManager(this)))
+{
+}
+
+///
+/// Constructor
+///
+/// \param netManager shared network access manager, needs to have at least the lifetime of this
+/// manager.
+///
+QXmppHttpUploadManager::QXmppHttpUploadManager(QNetworkAccessManager *netManager)
+    : d(std::make_unique<QXmppHttpUploadManagerPrivate>(netManager))
+{
+}
+
+QXmppHttpUploadManager::~QXmppHttpUploadManager() = default;
+
+///
+/// Upload the data from a QIODevice.
+///
+/// \param data QIODevice to read the data from. This can for example be a QFile.
```

Comment:
```suggestion
/// \param data QIODevice to read the data from. This can for example be a QFile.
///        It can be sequential or non-sequential.
```